include/range/v3/view/generate_n.hpp

```diff
@@ -67,7 +67,10 @@ namespace ranges
             {
                 RANGES_EXPECT(0 != rng_->n_);
                 --rng_->n_;
-                rng_->val_.reset();
+                if (rng_->val_)
+                    rng_->val_.reset();
+                else
+                    rng_->gen_();
```

Comment:
Hmmm...should we update `n_` after calling `gen_` for exception safety?

Reply:
I don't think it's hugely important, but it does seem more consistent with the behavior of `*i; ++i;` and `*i++` - so why not.